### Zukhra Mazhitova

There is a dataset with 100 sample of for-sale apartments in Atyrau city from Krisha.kz website. The task is about predicting the price of these apartments, model used to test is Logistic Regression, so it has to predict if the price of the apartment is more than 30 Million Tenges or not. 


Preprocessing methods used and how they are implemented

+ Data Cleaning 
    + Manually

+ Creating dummy variables
    + Functions
+ Normalization
    + Functions

+ Feature Selection
    + Functions


#### Data description

Data scrapped from Krisha.kz contains 100 records and 9 columns from the first 5 searched pages. 
+ id: a unique number of each apartment

+ title: the title of the apartment in the website

+ price: the price of the selling apartment

+ city: the city where apartment is located 
        
+ number: the contacting phone number of the seller

+ owner: company or individual owner of the apartment or posting the advertisement

+ district: district or region where apartment is located

+ square: area of the apartment, in m²(metres squared)

+ address: physical address, street or location of the apartment

In [2]:
#importing libraries
import pandas as pd
import numpy as np
import re

import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df = pd.read_excel (r'aparts2.xlsx',index_col='Unnamed: 0')
df.head()

title         price    city  \
0  1-комнатная квартира, 43.46 м², Абылхаир хана ...  15 646 500 〒  Атырау   
1  2-комнатная квартира, 65.53 м², Абылхаир хана ...  22 167 108 〒  Атырау   
2  1-комнатная квартира, 56.9 м², Авангард 2 микр...  15 932 000 〒  Атырау   
3  4-комнатная квартира, 120 м², 2/9 этаж, Жарбос...  44 000 000 〒  Атырау   
4  4-комнатная квартира, 136.5 м², 2-ая Береговая...  43 680 000 〒  Атырау   

            number                    owner          district     square  \
0  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   43.46 м²   
1  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   65.53 м²   
2  +7 778 017 1133  Застройщик TOO "ORKALI"   Атырауская обл.    56.9 м²   
3  +7 777 311 7000      Хозяин недвижимости   Атырауская обл.     120 м²   
4  +7 778 017 1153  Застройщик TOO "ORKALI"   Атырауская обл.   136.5 м²   

                       addresss  
0   Абылхаир хана — Жумагалиева  
1   Абылхаир хана — Жумагалиева  
2      Авангард 2 микрорайон 6А  
3                Жарбосынова 62  
4          2-ая Береговая линия

In [5]:
df['room']=df.title.str[0]
df.head()

title         price    city  \
0  1-комнатная квартира, 43.46 м², Абылхаир хана ...  15 646 500 〒  Атырау   
1  2-комнатная квартира, 65.53 м², Абылхаир хана ...  22 167 108 〒  Атырау   
2  1-комнатная квартира, 56.9 м², Авангард 2 микр...  15 932 000 〒  Атырау   
3  4-комнатная квартира, 120 м², 2/9 этаж, Жарбос...  44 000 000 〒  Атырау   
4  4-комнатная квартира, 136.5 м², 2-ая Береговая...  43 680 000 〒  Атырау   

            number                    owner          district     square  \
0  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   43.46 м²   
1  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   65.53 м²   
2  +7 778 017 1133  Застройщик TOO "ORKALI"   Атырауская обл.    56.9 м²   
3  +7 777 311 7000      Хозяин недвижимости   Атырауская обл.     120 м²   
4  +7 778 017 1153  Застройщик TOO "ORKALI"   Атырауская обл.   136.5 м²   

                       addresss room  
0   Абылхаир хана — Жумагалиева    1  
1   Абылхаир хана — Жумагалиева    2  
2      Авангард 2 микрорайон 6А    1  
3                Жарбосынова 62    4  
4          2-ая Береговая линия    4

In [4]:
df['room']=df['room'].astype(int)
df['price']=df['price'].str.strip(' 〒')
df['cost'] = df['price'].str.replace('\xa0', '')

df['cost']=df['cost'].astype(str).astype(int)

df['square']=df['square'].str.strip(' м²')
df['square']=df['square'].astype(float)


In [5]:
df.head()

Unnamed: 0                                              title        price  \
0           0  1-комнатная квартира, 43.46 м², Абылхаир хана ...  15 646 500    
1           1  2-комнатная квартира, 65.53 м², Абылхаир хана ...  22 167 108    
2           2  1-комнатная квартира, 56.9 м², Авангард 2 микр...  15 932 000    
3           3  4-комнатная квартира, 120 м², 2/9 этаж, Жарбос...  44 000 000    
4           4  4-комнатная квартира, 136.5 м², 2-ая Береговая...  43 680 000    

     city           number                    owner          district  square  \
0  Атырау  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   43.46   
1  Атырау  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   65.53   
2  Атырау  +7 778 017 1133  Застройщик TOO "ORKALI"   Атырауская обл.   56.90   
3  Атырау  +7 777 311 7000      Хозяин недвижимости   Атырауская обл.  120.00   
4  Атырау  +7 778 017 1153  Застройщик TOO "ORKALI"   Атырауская обл.  136.50   

                       addresss  room      cost  
0   Абылхаир хана — Жумагалиева     1  15646500  
1   Абылхаир хана — Жумагалиева     2  22167108  
2      Авангард 2 микрорайон 6А     1  15932000  
3                Жарбосынова 62     4  44000000  
4          2-ая Береговая линия     4  43680000

In [6]:
df_without = df
df['price'] = [ 'more 30M' if x > 30000000
                  else 'less'
                  for x in df['cost']]



In [7]:
df['price'] = [0 if x == 'less' else 1 for x in df['price']]


In [8]:
y = df['price']
df=df.drop(['cost','price'],1)
y


0     0
1     0
2     0
3     1
4     1
     ..
95    0
96    0
97    0
98    0
99    1
Name: price, Length: 100, dtype: int64

### Dummies

In [9]:
print(df['owner'].head(5))

0        Застройщик BI Group
1        Застройщик BI Group
2    Застройщик TOO "ORKALI"
3        Хозяин недвижимости
4    Застройщик TOO "ORKALI"
Name: owner, dtype: object


In [10]:
for col_name in df.columns:
    if df[col_name].dtypes == 'object':
        unique_cat = len(df[col_name].unique())
        print("Column '{col_name}' has {unique_cat} unique values".format(
            col_name=col_name, unique_cat=unique_cat
        ))

Column 'title' has 100 unique values
Column 'city' has 1 unique values
Column 'number' has 71 unique values
Column 'owner' has 13 unique values
Column 'district' has 15 unique values
Column 'addresss' has 80 unique values


In [11]:
print(df['owner'].value_counts().sort_values(ascending=False))

Хозяин недвижимости                    59
Кадирбаева Айзада Туралбековна         11
Застройщик TOO "ORKALI"                 8
Застройщик BI Group                     5
Сайназаров Сагынбек                     4
Некрасова Татьяна                       3
Мендыбаева Зарина                       2
Специалист Айгуль                       2
Агентство недвижимости Метры Атырау     2
Adresa                                  1
"А-Курман-Құрылыс Групп"                1
Гульсара Сайлауовна                     1
Калаганова Сауле                        1
Name: owner, dtype: int64


In [12]:
df['owner_cat'] = [ 'homeowner' if x == 'Хозяин недвижимости'
                  else 'builder' if x == '"А-Курман-Құрылыс Групп"' or 
                   x == 'Застройщик BI Group' or 
                   x == 'Застройщик TOO "ORKALI"' 
                  else 'realtor' if x == 'Агентство недвижимости Метры Атырау' or 
                  x == 'Гульсара Сайлауовна' or x == 'Кадирбаева Айзада Туралбековна' or
                  x == 'Калаганова Сауле' or x ==  'Мендыбаева Зарина' or x == 'Некрасова Татьяна'
                   or x=='Сайназаров Сагынбек'  or x == 'Специалист Айгуль'
                  else 'Other'
                  for x in df['owner']]
print(df['owner_cat'].value_counts().sort_values(ascending=False))

homeowner    59
realtor      26
builder      14
Other         1
Name: owner_cat, dtype: int64


In [13]:
print(pd.get_dummies(df['owner_cat']).head(5))

   Other  builder  homeowner  realtor
0      0        1          0        0
1      0        1          0        0
2      0        1          0        0
3      0        0          1        0
4      0        1          0        0


In [14]:
print(df['district'].value_counts().sort_values(ascending=False))

 Атырауская обл.     36
 мкр Нурсая          28
 Жилгородок           7
 Авангард-4           5
 Авангард-2           4
 мкр. Алмагуль        4
 мкр Центральный      3
 Ардагер              2
 Привокзальный-3А     2
 мкр Самал            2
 мкр Лесхоз           2
 Привокзальный-3      2
 Привокзальный-5      1
 пгт Балыкши          1
 Авангард-3           1
Name: district, dtype: int64


In [15]:
df['district_cat'] = [ 'Нурсая' if re.compile(r'Нурсая').search(x)
                  else 'Жилгородок' if re.compile(r'Жилгородок').search(x)
                  else 'Алмагуль' if re.compile(r'Алмагуль').search(x)
                  else 'Авангард' if re.compile(r'Авангард').search(x)
                  else 'Центральный'if re.compile(r'Центральный').search(x)
                  else 'Самал' if re.compile(r'Самал').search(x)
                  else 'Привокзальный'if re.compile(r'Привокзальный').search(x)
                  else 'Ардагер' if re.compile(r'Ардагер').search(x)
                  else 'Балыкши' if re.compile(r'Балыкши').search(x)
                  else 'Other'
                  for x in df['title']]
print(df['district_cat'].value_counts().sort_values(ascending=False))

Other            34
Нурсая           28
Авангард         14
Жилгородок        7
Привокзальный     5
Алмагуль          4
Центральный       3
Ардагер           2
Самал             2
Балыкши           1
Name: district_cat, dtype: int64


In [16]:
df['coast'] = [ 'europe' if re.compile('Абулхаир|Абылхаир|Жарбосынова|Береговая|Нурсая|Сатпаева|Владимирского|Ардагер|Ерниязова|Студенческий|Лесхоз|Самал|Тайманова|Курмангазы|Жарбосынова|Авангард|Кулманов').search(x)
                  else 'asia' if re.compile('Жилгородок|Датов|Чехова|Валиханова|Алмагуль|Азаттык|Атамбаева|Центральный|Балыкши|Привокзальный|Баймуханов|Утемисов|Азаттық').search(x)
                  else 'Other'
                  for x in df['title']]
print(df['coast'].value_counts().sort_values(ascending=False))

europe    71
asia      29
Name: coast, dtype: int64


In [17]:
dummy_column=['coast','district_cat','owner_cat']

In [18]:
def toDummy(df, dummy_column):
    for x in dummy_column:
        dummies=pd.get_dummies(df[x],prefix=x,dummy_na=False)
        df = df.drop(x,1)
        df = pd.concat([df, dummies], axis=1)
    return df

In [19]:
df = toDummy(df, dummy_column)
print(df.head(5))

   Unnamed: 0                                              title    city  \
0           0  1-комнатная квартира, 43.46 м², Абылхаир хана ...  Атырау   
1           1  2-комнатная квартира, 65.53 м², Абылхаир хана ...  Атырау   
2           2  1-комнатная квартира, 56.9 м², Авангард 2 микр...  Атырау   
3           3  4-комнатная квартира, 120 м², 2/9 этаж, Жарбос...  Атырау   
4           4  4-комнатная квартира, 136.5 м², 2-ая Береговая...  Атырау   

            number                    owner          district  square  \
0  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   43.46   
1  +7 778 017 1149      Застройщик BI Group   Атырауская обл.   65.53   
2  +7 778 017 1133  Застройщик TOO "ORKALI"   Атырауская обл.   56.90   
3  +7 777 311 7000      Хозяин недвижимости   Атырауская обл.  120.00   
4  +7 778 017 1153  Застройщик TOO "ORKALI"   Атырауская обл.  136.50   

                       addresss  room  coast_asia  ...  district_cat_Балыкши  \
0   Абылхаир хана — Жума

### Dropping unnessary columns

In [20]:
df = df.drop(['title','addresss','city','number','owner','district'], axis = 1)
print(df.head(5))

   Unnamed: 0  square  room  coast_asia  coast_europe  district_cat_Other  \
0           0   43.46     1           0             1                   1   
1           1   65.53     2           0             1                   1   
2           2   56.90     1           0             1                   0   
3           3  120.00     4           0             1                   1   
4           4  136.50     4           0             1                   1   

   district_cat_Авангард  district_cat_Алмагуль  district_cat_Ардагер  \
0                      0                      0                     0   
1                      0                      0                     0   
2                      1                      0                     0   
3                      0                      0                     0   
4                      0                      0                     0   

   district_cat_Балыкши  district_cat_Жилгородок  district_cat_Нурсая  \
0                     0  

In [21]:
df.head()

Unnamed: 0  square  room  coast_asia  coast_europe  district_cat_Other  \
0           0   43.46     1           0             1                   1   
1           1   65.53     2           0             1                   1   
2           2   56.90     1           0             1                   0   
3           3  120.00     4           0             1                   1   
4           4  136.50     4           0             1                   1   

   district_cat_Авангард  district_cat_Алмагуль  district_cat_Ардагер  \
0                      0                      0                     0   
1                      0                      0                     0   
2                      1                      0                     0   
3                      0                      0                     0   
4                      0                      0                     0   

   district_cat_Балыкши  district_cat_Жилгородок  district_cat_Нурсая  \
0                     0                        0                    0   
1                     0                        0                    0   
2                     0                        0                    0   
3                     0                        0                    0   
4                     0                        0                    0   

   district_cat_Привокзальный  district_cat_Самал  district_cat_Центральный  \
0                           0                   0                         0   
1                           0                   0                         0   
2                           0                   0                         0   
3                           0                   0                         0   
4                           0                   0                         0   

   owner_cat_Other  owner_cat_builder  owner_cat_homeowner  owner_cat_realtor  
0                0                  1                    0                  0  
1                0                  1                    0                  0  
2                0                  1                    0                  0  
3                0                  0                    1                  0  
4                0                  1                    0                  0

### Missing values

In [22]:
 df.isnull().sum().sort_values(ascending=False).head()

owner_cat_realtor       0
district_cat_Ардагер    0
square                  0
room                    0
coast_asia              0
dtype: int64

### Creating combinations of values

In [23]:
from itertools import combinations
from sklearn.preprocessing import PolynomialFeatures

combos = list(combinations(list(df.columns),2))
colnames = list(df.columns)+['_'.join(x) for x in combos]
poly = PolynomialFeatures(interaction_only=True, include_bias=False)
df=poly.fit_transform(df)
df=pd.DataFrame(df)
df.columns=colnames
noint_indices = [i for i, x in enumerate(list((df==0).all())) if x]
df=df.drop(df.columns[noint_indices],axis=1)

print(df.head())

   Unnamed: 0  square  room  coast_asia  coast_europe  district_cat_Other  \
0         0.0   43.46   1.0         0.0           1.0                 1.0   
1         1.0   65.53   2.0         0.0           1.0                 1.0   
2         2.0   56.90   1.0         0.0           1.0                 0.0   
3         3.0  120.00   4.0         0.0           1.0                 1.0   
4         4.0  136.50   4.0         0.0           1.0                 1.0   

   district_cat_Авангард  district_cat_Алмагуль  district_cat_Ардагер  \
0                    0.0                    0.0                   0.0   
1                    0.0                    0.0                   0.0   
2                    1.0                    0.0                   0.0   
3                    0.0                    0.0                   0.0   
4                    0.0                    0.0                   0.0   

   district_cat_Балыкши  ...  district_cat_Балыкши_owner_cat_realtor  \
0                   0.0  .

### Normalizing

In [24]:
from sklearn import preprocessing

normalized = pd.DataFrame(preprocessing.normalize(df.square.values.reshape(1,-1)))

normalized=normalized.transpose()
df['square_norm']= normalized
df=df.drop('square',axis=1)
df

Unnamed: 0  room  coast_asia  coast_europe  district_cat_Other  \
0          0.0   1.0         0.0           1.0                 1.0   
1          1.0   2.0         0.0           1.0                 1.0   
2          2.0   1.0         0.0           1.0                 0.0   
3          3.0   4.0         0.0           1.0                 1.0   
4          4.0   4.0         0.0           1.0                 1.0   
..         ...   ...         ...           ...                 ...   
95        95.0   2.0         0.0           1.0                 0.0   
96        96.0   3.0         0.0           1.0                 0.0   
97        97.0   2.0         0.0           1.0                 0.0   
98        98.0   2.0         1.0           0.0                 0.0   
99        99.0   2.0         0.0           1.0                 0.0   

    district_cat_Авангард  district_cat_Алмагуль  district_cat_Ардагер  \
0                     0.0                    0.0                   0.0   
1                     0.0                    0.0                   0.0   
2                     1.0                    0.0                   0.0   
3                     0.0                    0.0                   0.0   
4                     0.0                    0.0                   0.0   
..                    ...                    ...                   ...   
95                    0.0                    0.0                   0.0   
96                    0.0                    0.0                   0.0   
97                    0.0                    0.0                   0.0   
98                    0.0                    0.0                   0.0   
99                    0.0                    0.0                   0.0   

    district_cat_Балыкши  district_cat_Жилгородок  ...  \
0                    0.0                      0.0  ...   
1                    0.0                      0.0  ...   
2                    0.0                      0.0  ...   
3                    0.0                      0.0  ...   
4                    0.0                      0.0  ...   
..                   ...                      ...  ...   
95                   0.0                      0.0  ...   
96                   0.0                      0.0  ...   
97                   0.0                      0.0  ...   
98                   0.0                      1.0  ...   
99                   0.0                      0.0  ...   

    district_cat_Жилгородок_owner_cat_homeowner  \
0                                           0.0   
1                                           0.0   
2                                           0.0   
3                                           0.0   
4                                           0.0   
..                                          ...   
95                                          0.0   
96                                          0.0   
97                                          0.0   
98                                          1.0   
99                                          0.0   

    district_cat_Жилгородок_owner_cat_realtor  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         0.0   
4                                         0.0   
..                                        ...   
95                                        0.0   
96                                        0.0   
97                                        0.0   
98                                        0.0   
99                                        0.0   

    district_cat_Нурсая_owner_cat_builder  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     0.0   
..                                    ...   
95                                    0.0   
96             

In [25]:
X =df.drop('Unnamed: 0',1)

### Feature selection

In [26]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y,train_size=0.70, random_state=1)

In [27]:
import sklearn.feature_selection

select = sklearn.feature_selection.SelectKBest(k=20)
selected_features = select.fit(X_train, y_train)
indices_selected = selected_features.get_support(indices=True)
colnames_selected = [X.columns[i] for i in indices_selected]

X_train_selected = X_train[colnames_selected]
X_test_selected = X_test[colnames_selected]

C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:114: UserWarning: Features [ 13  31  48  64  82  86  94  99 103] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx,
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\feature_selection\_univariate_selection.py:116: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


In [28]:
print(colnames_selected)


['room', 'coast_asia', 'coast_europe', 'district_cat_Other', 'owner_cat_builder', 'owner_cat_realtor', 'Unnamed: 0_coast_asia', 'Unnamed: 0_owner_cat_builder', 'Unnamed: 0_owner_cat_realtor', 'square_room', 'square_coast_europe', 'square_district_cat_Other', 'square_owner_cat_builder', 'room_coast_europe', 'room_district_cat_Other', 'room_owner_cat_builder', 'coast_europe_district_cat_Other', 'coast_europe_owner_cat_builder', 'district_cat_Other_owner_cat_builder', 'square_norm']


### Building Logistic Regression model with prepocessed data 

In [29]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

def find_model_perf(X_train, y_train, X_test, y_test):
    model = LogisticRegression()
    model.fit(X_train, y_train)
    y_hat = [x[1] for x in model.predict_proba(X_test)]
    auc = roc_auc_score(y_test, y_hat)
    
    return auc

In [30]:
auc_processed = find_model_perf(X_train_selected, y_train, X_test_selected, y_test)
print(auc_processed)

0.8757763975155279


C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### Building Logistic Regression model without prepocessing 

In [31]:
for col_name in df_without.columns:
    if df_without[col_name].dtypes not in ['int32','int64','float32','float64']:
        df_without = df_without.drop(col_name, 1)

In [32]:
X_without = df_without.drop(['cost','Unnamed: 0','price'], 1)
y_without = y
X_without.head()

square  room
0   43.46     1
1   65.53     2
2   56.90     1
3  120.00     4
4  136.50     4

In [33]:
y_without

0     0
1     0
2     0
3     1
4     1
     ..
95    0
96    0
97    0
98    0
99    1
Name: price, Length: 100, dtype: int64

In [34]:
print(X_without.shape)
print(y_without.shape)

(100, 2)
(100,)


In [35]:
X_train_without, X_test_without, y_train, y_test = train_test_split(
    X_without, y_without, train_size=0.70, random_state=1)
auc_unprocessed = find_model_perf(X_train_without, y_train, X_test_without, y_test)
print(auc_unprocessed)


0.84472049689441


### Conclusion

While predicting whether the apartment's cost is above 30 million tenge or not, with preproccessing performance of Area under the ROC Curve was 0.858, while without preproccesing it is 0.845. Therefore, we can say that preprocessing sligtly increased the performance of the model. 